File: patterns/01-design-patterns/01-creational/03-builder/go/builder.go
package builder

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type ChatRequest struct {
	Model     string
	Messages  []string
	MaxTokens int
	Stream    bool
}
type ChatRequestBuilder struct {
	model     string
	messages  []string
	maxTokens int
	stream    bool
}

func (b *ChatRequestBuilder) Configure(value map[string]any) *ChatRequestBuilder {
	b.model, _ = value["model"].(string)
	b.messages = nil
	for _, m := range value["messages"].([]any) {
		b.messages = append(b.messages, m.(string))
	}
	b.maxTokens = int(value["max_tokens"].(float64))
	b.stream, _ = value["stream"].(bool)
	return b
}
func (b *ChatRequestBuilder) Build() (ChatRequest, error) {
	if b.model == "" || len(b.messages) == 0 || b.maxTokens <= 0 {
		return ChatRequest{}, &PatternError{code: "invalid_request"}
	}
	result := ChatRequest{b.model, append([]string(nil), b.messages...), b.maxTokens, b.stream}
	*b = ChatRequestBuilder{}
	return result, nil
}
func Evaluate(input map[string]any) (any, error) {
	b := &ChatRequestBuilder{}
	requests := []map[string]any{}
	for _, raw := range input["builds"].([]any) {
		product, err := b.Configure(raw.(map[string]any)).Build()
		if err != nil {
			return nil, err
		}
		requests = append(requests, map[string]any{"model": product.Model, "messages": product.Messages, "max_tokens": product.MaxTokens, "stream": product.Stream})
	}
	return map[string]any{"requests": requests}, nil
}

File: patterns/01-design-patterns/01-creational/03-builder/python/builder.py
from dataclasses import asdict, dataclass

class PatternError(Exception):
    def __init__(self, code): super().__init__(code); self.code = code

@dataclass(frozen=True)
class ChatRequest:
    model: str; messages: tuple[str, ...]; max_tokens: int; stream: bool

class ChatRequestBuilder:
    def __init__(self): self.reset()
    def reset(self): self.model=""; self.messages=[]; self.max_tokens=0; self.stream=False; return self
    def configure(self, value):
        self.model=value.get("model", ""); self.messages=list(value.get("messages", [])); self.max_tokens=value.get("max_tokens", 0); self.stream=value.get("stream", False); return self
    def build(self):
        if not self.model or not self.messages or self.max_tokens <= 0: raise PatternError("invalid_request")
        product=ChatRequest(self.model, tuple(self.messages), self.max_tokens, self.stream); self.reset(); return product

def evaluate(input_data):
    builder=ChatRequestBuilder(); requests=[]
    for value in input_data.get("builds", []):
        product=builder.configure(value).build(); item=asdict(product); item["messages"]=list(product.messages); requests.append(item)
    return {"requests": requests}
